From 8e830599fe2b0fbdf6adf1b3718209bf8ffe5ca7 Mon Sep 17 00:00:00 2001
From: Adrian Ratiu <adrian.ratiu@collabora.com>
Date: Thu, 3 Jul 2025 13:00:20 +0000
Subject: [PATCH] containers/vector_buffer: silence memcpy warn/err

Newer clang builds with -Wnontrivial-memcall with newer
glibc v2.41 header improvements prevent calling memset
memcpy & co on non-trivial data types.

We add a static_cast<void*> as a workaround and a TODO
to revisit in the future for a more clean solution, if
possible (eg after C++ 26).

The error is quite verbose, so I've added it in the bug
report [1]. An abridged version:

/build/arm-generic/usr/include/libchrome/base/containers/vector_buffer.h:149:17: error: first argument in call to 'memcpy' is a pointer to non-trivially copyable type 'element_type' (aka 'std::tuple<bool>') [-Werror,-Wnontrivial-memcall]
  149 |       memcpy(to.data(), from.data(), to.size_bytes());
      |                 ^

Bug: 387993322, 432507886
Test: cros-workon-amd64-generic start libchrome; emerge-amd64-generic iioservice

Link: https://partnerissuetracker.corp.google.com/issues/387993322#comment34 [1]
Change-Id: I3e48c500171c851a59ad111a21bef41ab34bf24f
---
 base/containers/vector_buffer.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/base/containers/vector_buffer.h b/base/containers/vector_buffer.h
index 6e9d9d092ce2d..65890f62ba732 100644
--- a/base/containers/vector_buffer.h
+++ b/base/containers/vector_buffer.h
@@ -146,7 +146,8 @@ class VectorBuffer {
     if constexpr (is_trivially_copyable_or_relocatable<T>) {
       // We can't use span::copy_from() as it tries to call copy constructors,
       // and fails to compile on move-only trivially-relocatable types.
-      memcpy(to.data(), from.data(), to.size_bytes());
+      // TODO(https://crbug.com/432507886): find a way to remove the void* cast
+      memcpy(static_cast<void*>(to.data()), from.data(), to.size_bytes());
       // Destructors are skipped because they are trivial or should be elided in
       // trivial relocation (https://reviews.llvm.org/D114732).
     } else {
-- 
2.39.5

